cleverhans/future/torch/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,204 @@
+            relative importance of size of the perturbation and
+            confidence of classification. If binary_search_steps is
+            large, the initial constant is not important. A smaller
+            value of this constant gives lower distortion results.
+            Default is 1e-2.
+  :param binary_search_steps: (optional) int. The number of times we
+            perform binary search to find the optimal tradeoff-constant
+            between norm of the perturbation and confidence of the
+            classification. Default is 5.
+  :param max_iterations: (optional) int. The maximum number of
+            iterations. Setting this to a larger value will produce
+            lower distortion results. Using only a few iterations
+            requires a larger learning rate, and will produce larger
+            distortion results. Default is 1000.
+  """
+  def compare(pred, label, is_logits=False):
+    """
+    A helper function to compare prediction against a label.
+    Returns true if the attack is considered successful.
+
+    :param pred: can be either a 1D tensor of logits or a predicted
+            class (int).
+    :param label: int. A label to compare against.
+    :param is_logits: (optional) bool. If True, treat pred as an
+            array of logits. Default is False.
+    """
+
+    # Convert logits to predicted class if necessary
+    if is_logits:
+      pred_copy = pred.clone().detach()
+      pred_copy[label] += confidence * (-1) ** int(targeted)
+      pred = torch.argmax(pred_copy)
+
+    return pred == label if targeted else pred != label
+
+
+  if y is None:
+    # Using model predictions as ground truth to avoid label leaking
+    pred = model_fn(x)
+    y = torch.argmax(pred, 1)
+
+  # Initialize some values needed for binary search on const
+  lower_bound = [0.] * len(x)
+  upper_bound = [1e10] * len(x)
+  const = x.new_ones(len(x), 1) * initial_const
+
+  o_bestl2 = [1e10] * len(x)
+  o_bestscore = [-1.] * len(x)
+  x = torch.clamp(x, clip_min, clip_max)
+  ox = x.clone().detach() # save the original x
+  o_bestattack = x.clone().detach()
+
+  # Map images into the tanh-space
+  # TODO as of 01/06/2020, PyTorch does not natively support
+  # arctanh (see, e.g.,
+  # https://github.com/pytorch/pytorch/issues/10324).
+  # This particular implementation here is not numerically
+  # stable and should be substituted w/ PyTorch's native
+  # implementation when it comes out in the future
+  arctanh = lambda x: 0.5 * torch.log((1 + x) / (1 - x))
+  x = (x - clip_min) / (clip_max - clip_min)
+  x = torch.clamp(x, 0, 1)
+  x = x * 2 - 1
+  x = arctanh(x * .999999)
+
+  # Prepare some variables
+  modifier = torch.zeros_like(x, requires_grad=True)
+  y_onehot = y.new_zeros((len(y), n_classes), dtype=torch.float)
+  y_onehot[range(len(y)), y] = 1.
+
+  # Define loss functions and optimizer
+  f_fn = lambda real, other, targeted: torch.max(
+    (-1) ** int(targeted) * (real - other) + confidence,
```

Comment:
Maybe same change here to make it more clear what's going on.

Reply:
Will do. Thanks.